### PR TITLE
Introduce frame::system::Error::InvalidOrigin

### LINKED ISSUE #4409: Introduce general BadOrigin runtime error
Right now module use various ways for reporting bad origins, even though the origin recognition APIs are all reusable library code. Some report a `&str` (of various content), others report a local `BadOrigin`-like error type.

Instead, there should be a single `BadOrigin` error (e.g. defined in `frame_system`) that all `ensure_origin`/`try_origin`/`ensure_root`/`ensure_nobody`/`ensure_signed` functions and their ilk use. If modules want to customise the error to something local, then they can with `map_err`, but a simple `ensure_root(origin)?` should Just Work.

### PR DESCRIPTION
**Pull request by DarkEld3r**
Key points:
- `InvalidOrigin` has been added to `frame::system::Error`.
- `EnsureOrigin` trait has been moved from `primitives/runtime/src/traits.rs` to `frame::system`.

Closes #4409.

### CODE REVIEW

File: primitives/runtime/src/traits.rs [hunk 1]
```diff
@@ -134,28 +134,6 @@ impl<
 	}
 }
 
-/// An error type that indicates that the origin is invalid.
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs]: I think the trait should stay here and we introduce the `BadOrigin` type directly here.

We already have some inherent errors that are declared by `decl_error!`. I would add a new one `BadOrigin` that implements `From<BadOrigin>` that is defined in this file here.

Comment (bkchr) [reply primitives/runtime/src/traits.rs]: I have some upcoming changes in my mind for the error stuff, that would conflict in the way it is implemented currently.

File: frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -168,7 +168,7 @@ decl_module! {
 		/// # </weight>
 		#[weight = SimpleDispatchInfo::FixedOperational(100_000)]
 		fn reject_proposal(origin, #[compact] proposal_id: ProposalIndex) {
-			T::RejectOrigin::ensure_origin(origin).map_err(|e| Into::<&str>::into(e))?;
+			T::RejectOrigin::ensure_origin(origin).map_err(|e| e.as_str())?;
```
Comment (gavofyork) [inline frame/treasury/src/lib.rs]: why does it need to be `.as_str()`? can't the error be returned back directly?

Comment (niklasad1) [reply frame/treasury/src/lib.rs]: I think `ensure_origin` returns `system::Error` which needs to be converted to `treasury::Error` accordingly to work but I think `Into::<&str>::into(e)` should still work.

**Comment by bkchr** [comment]
Superseded by: https://github.com/paritytech/substrate/pull/4449

**Pull request closed without merging**